File: benchmarks/cases/fdid_hongkong.py
"""Cross-validation: Forward DiD against both of Li's released implementations.

Li's headline application uses a confidential retailer panel, but she released
MATLAB and R code and a public companion dataset -- the Hsiao, Ching and Wan
(2012) Hong Kong GDP panel. Both of her implementations are vendored and run
live: ``Fun_FDID.R`` under Rscript, and ``FDID_Matlab.m`` with ``FDID_newR2.m``
under Octave at ``benchmarks/octave/fdid_hongkong.m``. Nothing here is
transcribed from her readme.

Running both buys three things.

Her two implementations are checked against each other, which nobody else does.
They agree to exactly 0.0 on every quantity both compute.

The MATLAB script computes inference her R script does not: the standardized
ATT, the standard error, the two-sided p-value and the 95 percent interval, for
the forward and the conventional fit alike. Those are now pinned, so the case
covers the uncertainty and not only the point estimate.

And the Octave run reproduces every figure her readme prints, to every digit she
printed: ``ATT_FDID = 0.025405``, ``ATT_std_FDID = 5.4941``,
``p_value_forward_DID = 3.9274e-08``, ``CI_95_FDID = [0.016342, 0.034468]``,
``ATT_DID = 0.031721``, ``ATT_std_DID = 3.8647``, ``p_value_DID = 0.00011122``,
nine controls selected.

One deliberate divergence, decomposed rather than tolerated. Li hardcodes 1.96
as the normal quantile (her ``FDID_Matlab.m`` line 49); mlsynth uses
``norm.ppf(0.975) = 1.959963984540``. That shifts each interval endpoint by
``3.601546e-05 * se``, which is 1.665e-07 on the forward fit and 2.956e-07 on the
conventional one -- matching the observed gaps of 1.67e-07 and 2.96e-07. So the
endpoints are pinned against Li's formula evaluated at mlsynth's own quantile,
where they agree to floating point, and the gap to her 1.96 interval is pinned
separately as the predicted quantity. A tolerance wide enough to swallow it would
have hidden a real drift of the same size.

Before this case ran at full precision, ``did_from_mean`` rounded its whole
return dict on the way out and ``results_assembly`` read the rounded vectors into
the typed result, so the tolerances here were 5e-4 and 2e-3 -- sized by mlsynth's
own display convention and nothing else. Li's code rounds nothing.
"""
from __future__ import annotations

from pathlib import Path

import pandas as pd

from mlsynth import FDID

from benchmarks.reference import reference_value

# basedata/HongKong.csv lives at the repo root.
_DATA = Path(__file__).resolve().parents[2] / "basedata" / "HongKong.csv"


def _fit():
    df = pd.read_csv(_DATA)
    return FDID(
        {
            "df": df,
            "outcome": "GDP",
            "treat": "Integration",
            "unitid": "Country",
            "time": "Time",
            "display_graphs": False,
            "verbose": False,
        }
    ).fit()


Z_EXACT = 1.959963984540054   # scipy.stats.norm.ppf(0.975)
Z_LI = 1.96                   # her FDID_Matlab.m lines 49-50 and 77-78


def run() -> dict:
    res = _fit()
    f, d = res.fdid, res.did
    r = lambda k: reference_value("fdid_hongkong", k)

    out = {
        # Li's own two implementations against each other.
        "li_r_vs_matlab_max_abs_diff": max(
            abs(r(k) - r("m_" + k)) for k in
            ("fdid_att", "fdid_att_pct", "fdid_r2_pre", "fdid_n_controls",
             "did_att", "did_att_pct", "did_r2_pre")),
        # mlsynth against MATLAB, on everything both compute the same way.
        "fdid_att": float(f.att),
        "fdid_att_pct": float(f.att_percent),
        "fdid_r2_pre": float(f.r_squared),
        "fdid_n_controls": float(len(f.selected_names)),
        "fdid_satt": float(f.satt),
        "fdid_se": float(f.att_se),
        "fdid_p_value": float(f.p_value),
        "did_att": float(d.att),
        "did_att_pct": float(d.att_percent),
        "did_r2_pre": float(d.r_squared),
        "did_satt": float(d.satt),
        "did_se": float(d.att_se),
        "did_p_value": float(d.p_value),
    }

    # The interval, decomposed. First: mlsynth's endpoints against Li's formula
    # evaluated at mlsynth's own quantile, which is the like-for-like check.
    out["fdid_ci_residual_at_our_z"] = max(
        abs(float(f.ci[0]) - (float(f.att) - Z_EXACT * float(f.att_se))),
        abs(float(f.ci[1]) - (float(f.att) + Z_EXACT * float(f.att_se))))
    out["did_ci_residual_at_our_z"] = max(
        abs(float(d.ci[0]) - (float(d.att) - Z_EXACT * float(d.att_se))),
        abs(float(d.ci[1]) - (float(d.att) + Z_EXACT * float(d.att_se))))
    # Second: the gap to her 1.96 interval is (1.96 - z) * se, and nothing else.
    for tag, fit in (("fdid", f), ("did", d)):
        predicted = (Z_LI - Z_EXACT) * float(fit.att_se)
        observed = abs(float(fit.ci[0]) - r(f"m_{tag}_ci_lo"))
        out[f"{tag}_ci_gap_is_the_z_difference"] = abs(observed - predicted)
    return out


def comparison() -> dict:
    """mlsynth ``FDID`` against Li's MATLAB, quantity by quantity.

    Pairs the point estimates and the inference -- the standardized ATT, the
    standard error, the p-value and both interval endpoints -- for the forward
    and the conventional fit. The MATLAB side is the live captured run at
    ``benchmarks/reference/fdid_hongkong/``; the endpoints differ by her
    hardcoded 1.96 against the exact quantile, which the case pins separately.
    """
    res = _fit()
    f, d = res.fdid, res.did
    pairs = [
        ("FDID/ATT", float(f.att), "m_fdid_att"),
        ("FDID/%ATT", float(f.att_percent), "m_fdid_att_pct"),
        ("FDID/R2_pre", float(f.r_squared), "m_fdid_r2_pre"),
        ("FDID/n_controls", float(len(f.selected_names)), "m_fdid_n_controls"),
        ("FDID/SATT", float(f.satt), "m_fdid_satt"),
        ("FDID/SE", float(f.att_se), "m_fdid_se"),
        ("FDID/p_value", float(f.p_value), "m_fdid_p_value"),
        ("FDID/CI_lo", float(f.ci[0]), "m_fdid_ci_lo"),
        ("FDID/CI_hi", float(f.ci[1]), "m_fdid_ci_hi"),
        ("DID/ATT", float(d.att), "m_did_att"),
        ("DID/%ATT", float(d.att_percent), "m_did_att_pct"),
        ("DID/R2_pre", float(d.r_squared), "m_did_r2_pre"),
        ("DID/SATT", float(d.satt), "m_did_satt"),
        ("DID/SE", float(d.att_se), "m_did_se"),
        ("DID/p_value", float(d.p_value), "m_did_p_value"),
        ("DID/CI_lo", float(d.ci[0]), "m_did_ci_lo"),
        ("DID/CI_hi", float(d.ci[1]), "m_did_ci_hi"),
    ]
    rows = [{"quantity": q, "mlsynth": v,
             "reference": reference_value("fdid_hongkong", k)}
            for q, v, k in pairs]
    cfg = {"outcome": "GDP", "treat": "Integration", "unitid": "Country",
           "time": "Time"}
    return {
        "rows": rows,
        "mlsynth_call": {"estimator": "FDID", "config": cfg},
        "reference": {
            "impl": ("Kathleen T. Li's FDID_Matlab.m + FDID_newR2.m under "
                     "Octave, and Fun_FDID.R under Rscript; both live runs"),
            "version": "Li (2024), Marketing Science, DOI 10.1287/mksc.2022.0212"},
    }


# Forward selection is deterministic, so both captured runs are exact re-runs,
# and the targets are read from the bundle rather than written here. The
# tolerances are what the implementations actually support: they used to be 5e-4
# and 2e-3 because mlsynth rounded its own output to 3-4 decimals on the way out,
# which is 7 to 9 orders of magnitude looser than the algorithms agree.
_fd = lambda k: reference_value("fdid_hongkong", k)
EXPECTED = {
    # Li's R and her MATLAB are bit-identical on every shared quantity.
    "li_r_vs_matlab_max_abs_diff": (0.0, 0.0),
    # The forward fit: measured at 5e-14 to 3e-13, pinned with BLAS headroom.
    "fdid_att": (_fd("m_fdid_att"), 1e-11),
    "fdid_att_pct": (_fd("m_fdid_att_pct"), 1e-11),
    "fdid_r2_pre": (_fd("m_fdid_r2_pre"), 1e-11),
    "fdid_n_controls": (_fd("m_fdid_n_controls"), 0.0),
    "fdid_satt": (_fd("m_fdid_satt"), 1e-11),
    "fdid_se": (_fd("m_fdid_se"), 1e-11),
    "fdid_p_value": (_fd("m_fdid_p_value"), 1e-11),
    # The conventional fit is a little looser: the donor average is summed in a
    # different order than R's and Octave's, so the ATT lands 1.9e-11 away and
    # the percentage amplifies that by 100/mean(yhat_post), about 3150x.
    "did_att": (_fd("m_did_att"), 1e-9),
    "did_att_pct": (_fd("m_did_att_pct"), 1e-6),
    "did_r2_pre": (_fd("m_did_r2_pre"), 1e-8),
    "did_satt": (_fd("m_did_satt"), 1e-8),
    "did_se": (_fd("m_did_se"), 1e-10),
    "did_p_value": (_fd("m_did_p_value"), 1e-11),
    # The interval agrees with Li's formula at our own quantile to floating point.
    "fdid_ci_residual_at_our_z": (0.0, 1e-15),
    "did_ci_residual_at_our_z": (0.0, 1e-15),
    # ...and the whole gap to her 1.96 interval is (1.96 - z) * se. Pinned at
    # zero residual: if the endpoints ever drift for any other reason, the
    # prediction stops accounting for the gap and this fails. The residual cannot
    # beat the ATT disagreement it sits on top of, so the conventional arm's
    # tolerance is set by its own 1.9e-11 summation difference and not by the
    # quantile arithmetic, which is exact.
    "fdid_ci_gap_is_the_z_difference": (0.0, 1e-12),
    "did_ci_gap_is_the_z_difference": (0.0, 1e-10),
}

File: benchmarks/octave/fdid_hongkong.m
% Kathleen T. Li's FDID_Matlab.m and FDID_newR2.m, from the Marketing Science
% replication package (DOI 10.1287/mksc.2022.0212), on her released Hong Kong
% GDP panel. Run under Octave so the benchmark can read the author's own MATLAB
% numbers live rather than transcribing her readme.
%
% The numerics are hers, line for line. Four changes, and nothing else:
%
%   1. normcdf -> 0.5*erfc(-x/sqrt(2)). normcdf lives in MATLAB's Statistics
%      Toolbox and in Octave's statistics package, which is not assumed here.
%      The substitution is the exact same function, not an approximation:
%      0.5*erfc(-1.96/sqrt(2)) = 0.975002104852.
%   2. The two plotting blocks (her lines 82-129) are dropped. They write PDFs
%      and need a display; the benchmark reads numbers.
%   3. The data path arrives as an argument instead of being the working
%      directory's GDP.csv.
%   4. Her FDID_newR2 is a separate file and stays one here, as fdid_newr2.m.
%
% One hazard of hers is left in place and flagged rather than fixed: at her
% FDID_newR2.m line 60, num_c = find(R2final == max(R2final)) returns every
% index attaining the maximum, where the R version's which.max returns the
% first. On a tie the two of her own implementations would diverge. On this
% panel the argmax is unique at 9, so it does not bite, and changing it would
% make this no longer her code.

% fdid_newr2.m and normcdf_.m sit next to this file, mirroring Li's own layout
% (she ships FDID_newR2.m separately), so put this directory on the path.
addpath(fileparts(mfilename('fullpath')));

args = argv();
datafile = 'GDP.csv';
for k = 1:numel(args)
  if strcmp(args{k}, '--data') && k < numel(args)
    datafile = args{k + 1};
  end
end

dataxy = csvread(datafile, 1, 0);   % her line 8: skip the country-name header
datay  = dataxy(:, 1);              % Hong Kong, the treated unit
datax  = dataxy(:, 2:end);          % the 24 controls

t  = size(datay, 1);                % 61
t1 = 44;                            % her line 14, the pre-treatment length
t2 = t - t1;                        % 17
y1 = datay(1:t1, 1);
y2 = datay(t1+1:t, 1);
x  = datax;
no_control = size(x, 2);

control_ID = [1:no_control];
[y1_hat_FDID, y2_hat_FDID, R2final, num_c, select_c] = ...
    fdid_newr2(no_control, control_ID, x, y1, y2, t1, t);

ATT_FDID       = mean(y2 - y2_hat_FDID);
ATT_FDID_per   = 100 * ATT_FDID / mean(y2_hat_FDID);
R2_forward_DID = 1 - (mean((y1 - y1_hat_FDID).^2)) / mean((y1 - mean(y1)).^2);

u1_FDID            = y1 - y1_hat_FDID;
Omega_1_hat_FDID   = (t2/t1) * mean(u1_FDID.^2);
Omega_2_hat_FDID   = mean(u1_FDID.^2);
std_Omega_hat_FDID = sqrt(Omega_1_hat_FDID + Omega_2_hat_FDID);

ATT_std_FDID        = sqrt(t2) * ATT_FDID / std_Omega_hat_FDID;
p_value_forward_DID = 2 * (1 - normcdf_(abs(ATT_std_FDID)));
CI_95_FDID_left     = ATT_FDID - 1.96 * std_Omega_hat_FDID / sqrt(t2);
CI_95_FDID_right    = ATT_FDID + 1.96 * std_Omega_hat_FDID / sqrt(t2);

% her lines 54-79: the conventional DID over all controls
x_DID   = mean(x, 2);
x1_DID  = x_DID(1:t1, :);
x2_DID  = x_DID(t1+1:t, :);
b_DID   = mean(y1 - x1_DID);
y1_DID  = b_DID + x1_DID;
y2_DID  = b_DID + x2_DID;

ATT_DID            = mean(y2 - y2_DID);
ATT_DID_percentage = 100 * ATT_DID / mean(y2_DID);
R2_DID             = 1 - (mean((y1 - y1_DID).^2)) / (mean((y1 - mean(y1)).^2));
u1_DID             = y1 - y1_DID;
Omega_1_hat_DID    = (t2/t1) * mean(u1_DID.^2);
Omega_2_hat_DID    = mean(u1_DID.^2);
std_Omega_hat_DID  = sqrt(Omega_1_hat_DID + Omega_2_hat_DID);
ATT_std_DID        = sqrt(t2) * ATT_DID / std_Omega_hat_DID;
p_value_DID        = 2 * (1 - normcdf_(abs(ATT_std_DID)));
CI_95_DID_left     = ATT_DID - 1.96 * std_Omega_hat_DID / sqrt(t2);
CI_95_DID_right    = ATT_DID + 1.96 * std_Omega_hat_DID / sqrt(t2);

printf('== MATLAB REFERENCE VALUES ==\n');
printf('m_fdid_att\t%.12f\n', ATT_FDID);
printf('m_fdid_att_pct\t%.12f\n', ATT_FDID_per);
printf('m_fdid_r2_pre\t%.12f\n', R2_forward_DID);
printf('m_fdid_n_controls\t%d\n', num_c);
printf('m_fdid_satt\t%.12f\n', ATT_std_FDID);
printf('m_fdid_se\t%.12f\n', std_Omega_hat_FDID / sqrt(t2));
printf('m_fdid_p_value\t%.12f\n', p_value_forward_DID);
printf('m_fdid_ci_lo\t%.12f\n', CI_95_FDID_left);
printf('m_fdid_ci_hi\t%.12f\n', CI_95_FDID_right);
printf('m_did_att\t%.12f\n', ATT_DID);
printf('m_did_att_pct\t%.12f\n', ATT_DID_percentage);
printf('m_did_r2_pre\t%.12f\n', R2_DID);
printf('m_did_satt\t%.12f\n', ATT_std_DID);
printf('m_did_se\t%.12f\n', std_Omega_hat_DID / sqrt(t2));
printf('m_did_p_value\t%.12f\n', p_value_DID);
printf('m_did_ci_lo\t%.12f\n', CI_95_DID_left);
printf('m_did_ci_hi\t%.12f\n', CI_95_DID_right);
printf('== MATLAB SESSION INFO ==\n');
printf('%s\n', version());
